core-v1/src/main/scala/data/Justified.scala

```diff
@@ -2,34 +2,139 @@ package com.rallyhealth.vapors.v1
+  def apply[V](value: V): Justified[V] = ByConst(value)
+
+  trait Visitor[G[+_]] {
+
+    def visitConstant[V](justified: ByConst[V]): G[V]
+
+    def visitConfig[V](justified: ByConfig[V]): G[V]
+
+    def visitFact[V](justified: ByFact[V]): G[V]
+
+    def visitInference[V](justified: ByInference[V]): G[V]
   }
 
-  final case class ByConfig[+V](
+  def byConst[V](value: V): Justified[V] = ByConst(value)
+
+  final case class ByConst[V](value: V) extends Justified[V] {
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitConstant(this)
+    override def evidence: NoEvidence.type = NoEvidence
+    override val reason: String = s"[const: $value]"
+    override def configs: Seq[(String, Option[String])] = Seq.empty
+    override def productPrefix: String = "Justified.ByConst"
+  }
+
+  def byConfig[V](
+    value: V,
+    configKey: String,
+    configDescription: Option[String] = None,
+  ): Justified[V] = ByConfig(value, configKey, configDescription)
+
+  final case class ByConfig[V](
     value: V,
     configKey: String,
     configDescription: Option[String] = None,
   ) extends Justified[V] {
-    override def withValue[U](value: U): Justified[U] = copy(value = value)
+    override def evidence: NoEvidence.type = NoEvidence
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitConfig(this)
+    override val reason: String = s"[config: $configKey=$value]"
+    override def configs: Seq[(String, Option[String])] = Seq(configKey -> configDescription)
+    override def productPrefix: String = "Justified.ByConfig"
   }
 
-  final case class ByEvidence[+V](
+  def byFact[V](fact: TypedFact[V]): Justified[V] = ByFact(fact)
+
+  // TODO: Allow a lens for map operations? Many valuesOfType operations will be followed directly by a select
+  //       operation. By using a lens here, we can easily chain map and select into a single operation to avoid
+  //       more deeply nested Justified trees and save space / time when storing / interpreting them.
+  // TODO: Alternatively, should there be a ByLens or BySelect justification that can chain multiple selects?
+  final case class ByFact[V](fact: TypedFact[V]) extends Justified[V] {
+    override val evidence: SomeEvidence = SomeEvidence(NonEmptySet.of(fact))
+    override def value: V = fact.value
+    override val reason: String = s"[fact: ${fact.typeInfo.name}=${fact.value}]"
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitFact(this)
+    override def configs: Seq[(String, Option[String])] = Seq.empty
+    override def productPrefix: String = "Justified.ByFact"
+  }
+
+  def byInference[V](
+    reason: String,
+    value: V,
+    sources: NonEmptyList[Justified[Any]],
+  ): Justified[V] = ByInference(reason, value, sources)
+
+  final case class ByInference[V](
+    reason: String,
     value: V,
-    evidence: Evidence,
+    sources: NonEmptyList[Justified[Any]], // TODO: Should this be a NonEmptySet? How to sort it?
```

Comment:
It seems like most operations have a specific order (like left & right, input & output, etc). Since we can't rely on type information, I don't think sorting is a good idea. I should remove this TODO